### DISCUSSION TITLE
Multitenancy support

### DISCUSSION THREAD

**Discussion by matze1610** [mikro-orm/mikro-orm#1821] [Q&A]
Hi Martin,
we are trying to use MikroORM for implementing a multitenancy application. We use schema separation for the tenant protection.

I know it's already on your roadmap - but is there a way we can support this with the existing board tools today already? The tenant Id  is carried by the Jwt token, so we would be able to create a fresh connection for each request. But would this make sense (perofrmance-wise and resource-wise)?
What would be your approach / recommendation? Can we help with the implementation?

What did you other guys out there do for multitenancy support?

Thanks
  Matze

**Comment by B4nan** [comment]
If you want to use multiple databases/connections, you should have multiple ORM instances. If it's a single server/user and multiple schemas, it should work fine with a single instance too. 

But can't really help much with this as I never worked with app using this approach, so I don't really know all the expectations and constraints.

**Reply by leonardbinet** [reply]
Hi, first congratulations for this library, I'm initiating the migration of an existing codebase from typeorm to mikro-orm and it's a relief to work with a well-documented/maintained ORM 🙏 

I'm trying to migrate our tenancy isolation using RLS from typeorm to mikro-orm and I encountered an issue linked to the request context it seems.

For a bit of context, I work in a company where we have a limited number of tenants, with lots of users in each tenant (B2B for large companies). Thus RLS makes sense (not too many connections, and easier setup).

Disclaimer: since I couldn't find the "proper" way to manage multiple orm instances, I've implemented the following pattern:

Here it is:
- all our tables have `tenant_id` column, and on each of those table a policy is applied that checks that filters rows based on following condition `CREATE POLICY "${tableName}_isolation_policy" ON "${tableName}" USING (FORMAT('tenant_%s', "${tenantIdColumn}"::TEXT) = current_user OR "${tenantIdColumn}" IS NULL)`
- thus each tenant has a dedicated orm instance using the `tenant_<tenantId>` as a username, which ensure tenant isolation a record level
- we also have an "admin" (non-tenant scoped connection) for which no policy applies, for multi-tenant operations operated by employees only (actually this instance is always instanciated first)
- to share connection among users of the same tenant, "tenant-scoped" orm instances are stored in singleton (stored as map `Map<TenantId, ORMInstance>`), 
- at each request, we identify the tenantId, and a (nestjs) request-scoped provider factory picks the orm instance for that tenant (or instanciate it if it doesn't exist), and forks it to ensure that multiple requests do not share the same identity map

In short, we bypass the  way request-context helper work, https://mikro-orm.io/docs/identity-map#how-does-requestcontext-helper-work, I actually don't even use the `MikroOrmModule.forRoot` detailed here https://docs.nestjs.com/recipes/mikroorm#request-scoped-handlers-in-queues.

I'm feeling I'm not doing it the "right way" but have difficulties understanding how it should be wired.

The problem I have is that when forking an entity manager instance, I get an entity manager with a different configuration:
<img width="1429" alt="Screenshot 2023-05-03 at 18 17 12" src="https://user-images.githubusercontent.com/1590371/235986779-4b2166a4-4ed1-48b6-b649-fb5f7b6aefb1.png">

In this case, when forking a tenant-scoped entity manager (`tenant_196`) I receive the "admin" one (`postgres_user`). Is this the expected behavior?

If necessary I can take time to provide a minimal repo for issue reproduction, but before doing this I was wondering if the issue was obvious to you?

Thank you,

Léonard

**Reply by leonardbinet** [reply]
Actually, reading https://github.com/mikro-orm/mikro-orm/issues/872 I found out `contextName` option which resolved my issue (using a `contextName` per orm instance).
I see this was implemented in https://github.com/mikro-orm/mikro-orm/commit/e11040d2b11ab97f2264f002908ea3dfcf514773 but wasn't described in documentation, it might be useful for other users using multiple ORM instances 🙂 

Any way, thanks again for this great lib 👏 .

**Comment by dginzbourg** [comment]
@matze1610 I'm thinking about going with the approach suggested in https://fabian.ski/posts/nestjs-tenants/. It works with Postgresql. The article also contains link to aws article for 3 different approaches how to handle multi-tenancy. IMO, row-based security gives the ability to move the security aspects to DB, keep your code written for a single tenant.

**Comment by yf-hk** [comment]
@dginzbourg In `v5` MikroORM will support `schema` option for all inserts, updates, queries etc. so it's possible to use PostgreSQL's schema to implement multi-tenancy now. IMO PostgreSQL's RLS isn't an ideal solution. When there are too many records it will greatly reduce performance and it's difficult to horizontally scale

**Reply by dginzbourg** [reply]
@dzcpy IMO, there're pros/cons for every solution. Having multiple DBs/schemas (1 per tenant) increases the management, e.g. schemas upgrade. It'd be interesting to hear if you have production that manages >1M tenants with schemas or DBs? we usually tend to go with 1 table with tenant column.
